src/dataflow/components/dataflow-program.tsx

```diff
@@ -494,7 +526,9 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
     const n1 = await nodeFactory!.createNode();
 
     const numNodes = this.programEditor.nodes.length;
-    n1.position = [5 + Math.floor((numNodes % 20) / 5) * 245 + Math.floor(numNodes / 20) * 15, 5 + numNodes % 5 * 90];
+    const { k } = this.programEditor.view.area.transform;
+    n1.position = [40 * (1 / k) + Math.floor((numNodes % 20) / 5) * 245 +
+                  Math.floor(numNodes / 20) * 15, 5 + numNodes % 5 * 90];
```

Comment:
There's a lot of magic numbers here and at first glance I have no idea what any of them are doing. I'd recommend a separate function named something like `getNewNodePosition()` which uses named constants where possible, e.g. `kNodesPerColumn = 5` and `kMargin = 5` (I think).

Reply:
Yeah, this got out of hand.  Refactored to use variables that should give a better sense of what is happening here.